### PR TITLE
Serato Tags: Add support for Track Color/BPM lock export

### PR DESCRIPTION
**Pull request by Holzhaus**

### CODE REVIEW

File: src/track/serato/markers2.cpp [hunk 1]
```diff
@@ -679,6 +679,22 @@ RgbColor::optional_t SeratoMarkers2::getTrackColor() const {
     return std::nullopt;
 }
 
+void SeratoMarkers2::setTrackColor(RgbColor color) {
+    for (auto& pEntry : m_entries) {
+        DEBUG_ASSERT(pEntry);
+        if (pEntry->typeId() != SeratoMarkers2Entry::TypeId::BpmLock) {
```
Comment (daschuer) [inline src/track/serato/markers2.cpp:685]: SeratoMarkers2Entry::TypeId::Color

Comment (daschuer) [reply src/track/serato/markers2.cpp:685]: Can you also add a test for that?

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you.

File: src/track/serato/markers2.cpp (hunk 1 above)
Comment (Holzhaus) [reply src/track/serato/markers2.cpp]: Fixed. I'll look into adding tests.

File: src/track/serato/markers2.cpp [hunk 2]
```diff
@@ -679,6 +679,22 @@ RgbColor::optional_t SeratoMarkers2::getTrackColor() const {
     return std::nullopt;
 }
 
+void SeratoMarkers2::setTrackColor(RgbColor color) {
+    for (auto& pEntry : m_entries) {
```
Comment (uklotzde) [inline src/track/serato/markers2.cpp:683]: const auto&, everywhere

File: src/track/serato/markers2.cpp [hunk 3]
```diff
@@ -679,6 +679,22 @@ RgbColor::optional_t SeratoMarkers2::getTrackColor() const {
...
+        DEBUG_ASSERT(pEntry);
+        if (pEntry->typeId() != SeratoMarkers2Entry::TypeId::Color) {
+            continue;
+        }
+        SeratoMarkers2ColorEntry* pColorEntry =
```
Comment (uklotzde) [inline src/track/serato/markers2.cpp:688]: http://www.cplusplus.com/reference/memory/static_pointer_cast/

File: src/track/serato/markers2.cpp [hunk 4]
```diff
@@ -679,6 +679,22 @@ RgbColor::optional_t SeratoMarkers2::getTrackColor() const {
...
+                static_cast<SeratoMarkers2ColorEntry*>(pEntry.get());
+        pColorEntry->setColor(color);
+        return;
+    }
+
+    SeratoMarkers2ColorEntry* pEntry = new SeratoMarkers2ColorEntry(color);
```
Comment (uklotzde) [inline src/track/serato/markers2.cpp:694]: We should use `make_shared` here, maybe followed by a static_pointer_cast if necessary. Maybe an *upcast* of `shared_ptr` does not require explicit casting?

File: src/track/serato/markers2.cpp [hunk 5]
```diff
@@ -695,4 +711,20 @@ bool SeratoMarkers2::isBpmLocked() const {
     return false;
 }
 
+void SeratoMarkers2::setBpmLocked(bool bpmLocked) {
+    for (auto& pEntry : m_entries) {
```
Comment (uklotzde) [inline src/track/serato/markers2.cpp:715]: For extra safety to verify your assumptions about the file format:

```
DEBUG_ASSERT(countEntriesByType(SeratoMarkers2Entry::TypeId::BpmLock) <= 1);
```

**Review by uklotzde — COMMENTED** [review]
Some minor suggestions

File: src/track/serato/markers2.cpp [hunk 6]
```diff
@@ -695,4 +711,20 @@ bool SeratoMarkers2::isBpmLocked() const {
...
+        DEBUG_ASSERT(pEntry);
+        if (pEntry->typeId() != SeratoMarkers2Entry::TypeId::BpmLock) {
+            continue;
+        }
+        SeratoMarkers2BpmLockEntry* pBpmLockEntry =
+                static_cast<SeratoMarkers2BpmLockEntry*>(pEntry.get());
+        pBpmLockEntry->setLocked(bpmLocked);
+        return;
```
Comment (uklotzde) [inline src/track/serato/markers2.cpp:723]: Add the assertion right here before the return statement. This is where it matters.

File: src/track/serato/markers2.cpp (hunk 4 above)
Comment (Holzhaus) [reply src/track/serato/markers2.cpp:694]: Oh right, good idea.

File: src/track/serato/markers2.cpp [hunk 7]
```diff
@@ -667,32 +670,97 @@ QByteArray SeratoMarkers2::dumpFLAC() const {
...
+        VERIFY_OR_DEBUG_ASSERT(pEntry) {
+            continue;
+        }
+
         if (pEntry->typeId() != SeratoMarkers2Entry::TypeId::Color) {
             continue;
         }
-        const SeratoMarkers2ColorEntry* pColorEntry = static_cast<SeratoMarkers2ColorEntry*>(pEntry.get());
+
+        const std::shared_ptr<SeratoMarkers2ColorEntry> pColorEntry =
```
Comment (uklotzde) [inline src/track/serato/markers2.cpp:682]: `auto` reduces the noise. The return type of static_pointer_cast is obvious.

**Review by daschuer — APPROVED** [review]
@uklotzde: merge?

**Review by uklotzde — APPROVED** [review]

**Comment by uklotzde** [comment]
Sure. LGTM.

**Pull request merged**